x-pack/plugins/apm/server/lib/service_map/get_service_map_from_trace_ids.test.ts

```diff
@@ -17,97 +31,146 @@ describe('getConnections', () => {
         typeof fetchServicePathsFromTraceIds
       >;
 
-      const connections = getConnections(
-        response.aggregations?.service_map.value.paths
-      );
-
-      expect(connections).toMatchSnapshot();
+      const connections = getConnections({
+        paths: response.aggregations?.service_map.value.paths,
+        serviceName: undefined,
+        environment: undefined,
+      });
+      const connectionsPairs = getConnectionsPairs(connections);
+      expect(connectionsPairs).toEqual([
+        'opbeans-go:testing -> opbeans-java:testing',
+        'opbeans-node:production -> opbeans-python:production',
+        'opbeans-node:production -> opbeans-ruby:production',
+        'opbeans-python:null -> opbeans-node:null',
+        'opbeans-python:production -> opbeans-node:production',
+        'opbeans-python:production -> opbeans-ruby:production',
+        'opbeans-ruby:production -> opbeans-node:production',
+        'opbeans-ruby:production -> opbeans-python:production',
+      ]);
     });
   });
 
   describe('if service name and environment are given', () => {
-    it('excludes connections with different service name and environment', () => {
+    it('shows all connections for opbeans-node and production', () => {
       const response = serviceMapFromTraceIdsScriptResponse as PromiseReturnType<
         typeof fetchServicePathsFromTraceIds
       >;
-      const serviceName = 'opbeans-node';
-      const environment = 'production';
 
-      const connections = getConnections(
-        response.aggregations?.service_map.value.paths,
-        serviceName,
-        environment
-      );
-      expect(
-        connections.every(
-          (conn) => conn.source['service.environment'] === environment
-        )
-      ).toBeTruthy();
-      const serviceNames = new Set();
-      connections.forEach((conn) =>
-        serviceNames.add(conn.source['service.name'])
-      );
-      ['opbeans-python', 'opbeans-node', 'opbeans-ruby'].forEach((name) =>
-        expect(serviceNames.has(name)).toBeTruthy()
-      );
+      const connections = getConnections({
+        paths: response.aggregations?.service_map.value.paths,
+        serviceName: 'opbeans-node',
+        environment: 'production',
+      });
+
+      const connectionsPairs = getConnectionsPairs(connections);
+
+      expect(connectionsPairs).toEqual([
+        'opbeans-node:production -> opbeans-python:production',
+        'opbeans-node:production -> opbeans-ruby:production',
+        'opbeans-python:production -> opbeans-node:production',
+        'opbeans-python:production -> opbeans-ruby:production',
+        'opbeans-ruby:production -> opbeans-node:production',
+        'opbeans-ruby:production -> opbeans-python:production',
+      ]);
+    });
```

Comment:
I personally split tests like these up in three parts:

- one that asserts whether all connection pairs include `opbeans-node:production`
- one that asserts whether no connection pairs include other environments (for `opbeans-node`) (the test data should have these connections to make the test useful)
- one that compares values

For the latter, I prefer using snapshots to more easily update this when we have a new data set (this applies to both API tests and unit tests).

Reply:
> - one that asserts whether all connection pairs include opbeans-node:production
> - one that asserts whether no connection pairs include other environments (for opbeans-node) (the test data should have these connections to make the test useful)

These two options are being covered in the assert I did in the current test, right? It asserts for the connections and environments.
FYI: the test data contains multiple environments.

Reply:
> - one that compares values

Do you mean a full snapshot of the data returned by `getConnections` function?

Reply:
From what I can tell you're just asserting whether the value returned from the test equals the array it currently returns. It's conceptually similar to a snapshot test. The mock response you have is huge - there's no way we can tell by looking at the input whether the output is correct. So, the next time someone updates the mock response, it's hard for that person to tell what exactly this was supposed to test. That's why I prefer some logical assertions next to an inline snapshot test. Right now you only have the latter (via `toEqual()`). Also, I do prefer `toMatchInlineSnapshot()` because it's just way more easy to automatically update snaphots.

Btw: in this case it might be even better to create a manual mock response, that is very small and intentful about the data it has. Then the values we expect can be reasoned about by looking at the mock response (hope that makes sense).

@sqren: would be interested in your input as well.

Reply:
>  That's why I prefer some logical assertions next to an inline snapshot test

I completely agree with this. In this case, I think this is the closest we get to a scoped down logical assertion. But you might have specific improvements in mind that we didn't think of.

Reply:
> one that asserts whether all connection pairs include `opbeans-node:production`

`connections` is a flat array, right? And it is expected to have other connections than `opbeans-node:production` because it calls other services that in turn will call other services. 

Might be better if we could verify that the root node was `opbeans-node:production` but I don't see how that's possible without calculating the full graph (which is another test entirely).

Reply:
My mistake, I mean paths rather than connections (a path is an array of successive connections). We should test whether if a path travels across `service.name:opbeans-node AND service.environment:production`, its connections (so those of the path) are included in the result. I don't see any test that does this. In fact, the values in the `toEqual` assertion tell me very little about what it is testing (without looking at a huge mock response). Thinking about it some more I feel like we should drop the approach where we use a JSON file to mock the response, and just inline a small set of paths to help us test all cases. I think coverage reports would help us discover what cases we need to test.

FWIW, don't want to bikeshed or block this PR for much longer, but it's a useful discussion for sure.

Reply:
I think it's valuable feedback. We went back and fourth a lot on how to best test this, and ended with something that felt decent.


> Thinking about it some more I feel like we should drop the approach where we use a JSON file to mock the response, and just inline a small set of paths to help us test all cases. I think coverage reports would help us discover what cases we need to test.

Not opposed to this but I'll leave it up to @cauemarcondes since he's already done some loops on this. I'm with either way.

Reply:
Aye. I do have an important question about checking for `span.type !== 'external'` (mentioning it here in case it gets missed).